haiku-apps/projectconceptor/projectconceptor-1.0.0.recipe

```diff
@@ -0,0 +1,50 @@
+SUMMARY="A Haiku diagramming application like Microsoft Visio"
+DESCRIPTION="ProjectConceptor is a modular application for Haiku and BeOS \
+designed to view, edit and process information which can be represented by a \
+graph structure."
+APP="ProjectConceptor"
+HOMEPAGE="https://github.com/Paradoxianer/$APP"
+COPYRIGHT="2017 Paradoxianer"
+LICENSE="MIT"
+REVISION="1"
+srcGitRev="63848b80dc520dc921b4be26e4f0fff8ec2155a9"
+SOURCE_URI="$HOMEPAGE/archive/$srcGitRev.tar.gz"
+CHECKSUM_SHA256="8a19f771f017b2ce90c36960c512c258bd0f571743b0bb00118577e10bdaa194"
+SOURCE_FILENAME="$APP-$portVersion-$srcGitRev.tar.gz"
+SOURCE_DIR="$APP-$srcGitRev"
+PATCHES="projectconceptor-1.0.0.patchset"
+
+ARCHITECTURES="x86_gcc2 !x86_64"
+
+PROVIDES="
+	projectconceptor = $portVersion
+	app:$APP = $portVersion
+	"
+REQUIRES="
+	haiku
+	"
+
+BUILD_REQUIRES="
+	haiku_devel
+	"
+BUILD_PREREQUIRES="
+	makefile_engine
+	cmd:g++
+	cmd:find
+	cmd:make
+	"
+
+BUILD()
+{
+	make -C src OBJ_DIR=objects $jobArgs
+}
+
+INSTALL()
+{
+	install -d $appsDir/Plugins
```

Comment:
The Plugins directory should not be created in `$appsDir` but in a subdirectory, e.g. `$appsDir/ProjectConceptor/Plugins/` or `$appsDir/ProjectConceptor-Plugins` or, even better, imho, in `$libDir/ProjectConceptor/Plugins/`.
Obviously, if you choose `$appsDir/ProjectConceptor/Plugins/` you will need to install the main runtime as `$appsDir/ProjectConceptor/ProjectConceptor` instead of `$appsDir/ProjectConceptor`.

Reply:
$appsDir/ProjectConceptor/add-ons/ is my preferred solution (with the executable in $appsDir/ProjectConceptor/ProjectConceptor).

$libDir and $addOndDir could be used, but I think they are relevant only for things that can be shared by multiple applications (eg. translators). In this case, the plugins are specific to the application so they should stay in the application directory.

(and in BeOS/Haiku terminology, we say "add-on", not "plugin").

Reply:
I have added a subdirectory ProjectConceptor and used "add-ons" in place of "Plugins" underneath it. Please see the updated commit below.

Reply:
I agree with some of your remarks, but:
* The terminology used inside an app should probably not be changed, because if ProjectConceptor's documentation mentions a "Plugins" directory then, imo, we should keep that name.
* There are several packages which create subdirectories in `/system/lib/` which are not used by any other recipes. I can think of a few: media-video/vlc, net-misc/openssh, app-misc/mc,
* `/system/lib/git/` is used by only 2 packages: git and git_flow.

For those reasons I would recommend using either `$appsDir/ProjectConceptor/Plugins/` or an better alternative imo: `$libDir/ProjectConceptor/Plugins/`. After all, if someone wanted to write a plugin for ProjectConceptor using another license, putting it in there would be more appropriate than putting it in `$appsDir/ProjectConceptor/add-ons/` or even `$appsDir/ProjectConceptor/Plugins/`.

Another advantage of using `$libDir/ProjectConceptor/Plugins/` is that the app can be installed directly in `$libDir`, which is nice because `/boot/system/apps` is in `$PATH`, so you can easily launch it from a terminal.

Reply:
Ok. I have changed it back to "Plugins". :)

Reply:
> There are several packages which create subdirectories in /system/lib/ which are not used by any other recipes. I can think of a few: media-video/vlc, net-misc/openssh, app-misc/mc,

If these libs are private to the application, then these should be fixed. If the libs are meant to be public, but just happens to not be used by anyone else, then it is ok. For example this is the case for NetSurf's libraries: they have an API and are documented separately from NetSurf, but not many projects use them (yet).

> Another advantage of using $libDir/ProjectConceptor/Plugins/ is that the app can be installed directly in $libDir, which is nice because /boot/system/apps is in $PATH, so you can easily launch it from a terminal.

First, plug-ins should be stored in $addOnsDir, not in $libDir. This is where load_add_on will look for them by default. There is not much technical difference between applications, libs and add-ons in Haiku (you could make a file that behaves well in all 3 cases). But we try to separate them cleanly.

The definitions are as follows:
- app: something you can run
- library: something you can link your application to. Usually an hard dependency (you cannot build the app without the lib)
- add-on: something optional that extends the app. App is not linked to it, instead it is loaded dynamically with load_add_on or dlopen. Can be added to an app after the fact.

Then there is the distinction of "public" vs "private" add-ons and libraries
- public: multiple apps can use the same library (libSDL, libbe, ...) or add-on (translators, media codecs, ...)
- private: only one application can use them

Here we are clearly in the private case. You may be right that these should go in a subdir of $addOnDir. This is the case for example for screensavers. So let's go for that (but in $addOnDir, not $libDir).

Reply:
Isn't `$addOnsDirs` reserved for stuff closely related to the system?
I consider that screensavers, as well as some very special exceptions like Qt5, are OK in `$addOnsDirs`.
But I still think that stuff like media-video/vlc, net-misc/openssh, app-misc/mc or app-crypt/johntheripper are fine with their subdirectories in `/system/lib/` (and should not be moved to `/system/add-ons/`.)

So I'd still recommend using `$libDir/ProjectConceptor/Plugins/` (and not `$addOnsDirs/ProjectConceptor/Plugins/`) for this recipe.

Reply:
I have a description of add-ons vs libraries in my message above. It is not a matter of public/private, but of dependency direction. An app *depends* on a library.  An app *uses* an add-on, but does not depend on it and can work without it.

UNIX apps (vlc, openssh, ...) does not usually have this distinction, and we have been lazy about patching them. But for native apps we should do the right thing.

Reply:
I do admit that ProjectConceptor's plugins are shared objects, but I'm not saying that they are libraries that other apps can use like they use "classical" libraries.

I claim that, in my opinion, putting these shared objects in `/system/lib/ProjectConceptor/Plugins/` is more appropriate than putting them in `/system/add-ons/ProjectConceptor/`.

If you are suggesting that we should move `/system/lib/vlc/`, `/system/lib/openssh/`, `/system/lib/mc/` and `/system/lib/john/` to `/system/add-ons/` then I disagree.

Reply:
OK, thanks for the explanation.
So, if I understand correctly, it is acceptable to create a subdirectory in `/system/lib/` and put ProjectConceptor's plugins there.

Would it also be acceptable to create one extra "Plugins" subdirectory in there, to have the plugins in `/system/lib/ProjectConceptor/Plugins/` ?
Thanks!

Reply:
The app is hardwired to the directory name "Plugins" and assumes that it's in the same directory where the app is. So "ProjectConceptor/Plugins" and "ProjectConceptor/ProjectConceptor" is the only way for now?